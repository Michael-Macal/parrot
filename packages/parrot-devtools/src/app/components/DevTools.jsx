import React, { Component } from 'react';
import PropTypes from 'prop-types';
import injectTapEventPlugin from 'react-tap-event-plugin';
import styled from 'styled-components';

// Material UI Components
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import Paper from 'material-ui/Paper';
import CircularProgress from 'material-ui/CircularProgress';

// Plugin Settings Components
import ListenerControls from './plugins/ListenerControls';
import GeneralControls from './plugins/GeneralControls';
import MiddlewareControls from './plugins/MiddlewareControls';

import fetchApi from '../utils/fetchApi';

// Needed for onTouchTap
// http://stackoverflow.com/a/34015469/988941
injectTapEventPlugin();

const pluginsConfig = {
  'parrot-middleware': {
    component: MiddlewareControls,
    label: 'Middleware Settings',
  },
  'parrot-listener': {
    component: ListenerControls,
    label: 'Listener Settings',
  },
};

const DevSection = styled.div`padding-top: 20px;`;
const DevHeader = styled.h5`margin: 0;`;

const paperStyle = {
  position: 'absolute',
  padding: '0 20',
  top: 10,
  bottom: 10,
  left: 10,
  right: 10,
};

class DevTools extends Component {
  static propTypes = {
    url: PropTypes.string,
  };

  static defaultProps = {
    url: 'http://localhost:3000',
  };

  state = {
    plugins: [],
    url:
      (window && window.localStorage && window.localStorage.getItem('parrotHostname')) ||
      this.props.url,
    error: false,
  };

  componentDidMount() {
    this.loadSections().catch(error => this.setState({ error }));
  }

  setUrl = url => {
    this.setState({ url });
    if (window && window.localStorage) {
      window.localStorage.setItem('parrotHostname', url);
    }
  };

  loadSections = async () => {
    // Empty cache
    this.setState({ error: false, plugins: [] });
    try {
      const resp = await fetchApi(this.state.url, '/parrot/registry');
      const registry = await resp.json();
      if (registry && registry.middlewares) {
        this.setState({ plugins: registry.middlewares });
      }
    } catch (error) {
      this.setState({ error });
    }
  };

  render() {
    let pluginSections;
    if (this.state.error) {
      pluginSections = (
        <DevSection>
          <h4>Squawk! Error alert!</h4>
          <p>
            Cannot find any Parrot plugins registered, can you check that Parrot is running on that
            hostname?
          </p>
        </DevSection>
      );
    } else if (this.state.plugins.length > 0) {
      pluginSections = this.state.plugins.map(pluginKey => pluginsConfig[pluginKey] || {}).map(
        ({ label, component: SectionComponent }) =>
          label && SectionComponent ? (
            <div>
              <DevSection>
                <DevHeader>{label}</DevHeader>
                <SectionComponent url={this.state.url} />
              </DevSection>
            </div>
          ) : null
      );
    } else {
      pluginSections = (
        <DevSection>
          <CircularProgress />
        </DevSection>
      );
    }

    return (
      <MuiThemeProvider>
        <Paper style={paperStyle}>
          {pluginSections}
          {/* General Settings are always shown */}
          <DevSection>
            <DevHeader>Parrot Settings</DevHeader>
            <GeneralControls
              url={this.state.url}
              setUrl={this.setUrl}
              onRefresh={this.loadSections}
            />
          </DevSection>
        </Paper>
      </MuiThemeProvider>
    );
  }
}

export default DevTools;
