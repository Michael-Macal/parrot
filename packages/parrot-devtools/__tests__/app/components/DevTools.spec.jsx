import React from 'react';
import { shallow } from 'enzyme';

import DevTools from '../../../src/app/components/DevTools';

jest.mock('../../../src/app/utils/fetchApi');

describe('DevTools component', () => {
  it('renders', () => {
    const wrapper = shallow(<DevTools />);
    expect(wrapper).toMatchSnapshot();
  });
});
