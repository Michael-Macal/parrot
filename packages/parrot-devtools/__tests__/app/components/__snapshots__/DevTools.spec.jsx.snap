// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DevTools component renders 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <MuiThemeProvider>
    <Paper
        circle={false}
        rounded={true}
        style={
            Object {
                "bottom": 10,
                "left": 10,
                "padding": "0 20",
                "position": "absolute",
                "right": 10,
                "top": 10,
              }
        }
        transitionEnabled={true}
        zDepth={1}
    >
        <styled.div>
            <CircularProgress
                max={100}
                min={0}
                mode="indeterminate"
                size={40}
                thickness={3.5}
                value={0}
            />
        </styled.div>
        <styled.div>
            <styled.h5>
                Parrot Settings
            </styled.h5>
            <GeneralControls
                onRefresh={[Function]}
                setUrl={[Function]}
                url="http://localhost:3000"
            />
        </styled.div>
    </Paper>
</MuiThemeProvider>,
  "nodes": Array [
    <MuiThemeProvider>
      <Paper
            circle={false}
            rounded={true}
            style={
                  Object {
                        "bottom": 10,
                        "left": 10,
                        "padding": "0 20",
                        "position": "absolute",
                        "right": 10,
                        "top": 10,
                      }
            }
            transitionEnabled={true}
            zDepth={1}
      >
            <styled.div>
                  <CircularProgress
                        max={100}
                        min={0}
                        mode="indeterminate"
                        size={40}
                        thickness={3.5}
                        value={0}
                  />
            </styled.div>
            <styled.div>
                  <styled.h5>
                        Parrot Settings
                  </styled.h5>
                  <GeneralControls
                        onRefresh={[Function]}
                        setUrl={[Function]}
                        url="http://localhost:3000"
                  />
            </styled.div>
      </Paper>
</MuiThemeProvider>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <DevTools
        url="http://localhost:3000"
/>,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": DevTools {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "loadSections": [Function],
        "props": Object {
          "url": "http://localhost:3000",
        },
        "refs": Object {},
        "setUrl": [Function],
        "state": Object {
          "error": false,
          "plugins": Array [],
          "url": "http://localhost:3000",
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <MuiThemeProvider>
          <Paper
                    circle={false}
                    rounded={true}
                    style={
                              Object {
                                        "bottom": 10,
                                        "left": 10,
                                        "padding": "0 20",
                                        "position": "absolute",
                                        "right": 10,
                                        "top": 10,
                                      }
                    }
                    transitionEnabled={true}
                    zDepth={1}
          >
                    <styled.div>
                              <CircularProgress
                                        max={100}
                                        min={0}
                                        mode="indeterminate"
                                        size={40}
                                        thickness={3.5}
                                        value={0}
                              />
                    </styled.div>
                    <styled.div>
                              <styled.h5>
                                        Parrot Settings
                              </styled.h5>
                              <GeneralControls
                                        onRefresh={[Function]}
                                        setUrl={[Function]}
                                        url="http://localhost:3000"
                              />
                    </styled.div>
          </Paper>
</MuiThemeProvider>,
        "_renderedOutput": <MuiThemeProvider>
          <Paper
                    circle={false}
                    rounded={true}
                    style={
                              Object {
                                        "bottom": 10,
                                        "left": 10,
                                        "padding": "0 20",
                                        "position": "absolute",
                                        "right": 10,
                                        "top": 10,
                                      }
                    }
                    transitionEnabled={true}
                    zDepth={1}
          >
                    <styled.div>
                              <CircularProgress
                                        max={100}
                                        min={0}
                                        mode="indeterminate"
                                        size={40}
                                        thickness={3.5}
                                        value={0}
                              />
                    </styled.div>
                    <styled.div>
                              <styled.h5>
                                        Parrot Settings
                              </styled.h5>
                              <GeneralControls
                                        onRefresh={[Function]}
                                        setUrl={[Function]}
                                        url="http://localhost:3000"
                              />
                    </styled.div>
          </Paper>
</MuiThemeProvider>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <DevTools
    url="http://localhost:3000"
/>,
}
`;
